File: designSearchAutoComplete.py
#Time Complexity : O(l*n + mlog3) where l is the average length of each sentence and n is the number of given sentences, and m is the average length of the given input substrings
#Space Complexity : O(l*n) where l is the average length of each sentence and n is the number of given sentences
#Did this code successfully run on Leetcode : Yes


class Pair:
    def __init__(self, sentence, count):
        self.sentence = sentence
        self.count = count

class TrieNode:
    def __init__(self):
        self.hashmap = {}
        self.children = {}

class AutocompleteSystem:

    def __init__(self, sentences: List[str], times: List[int]):
        self.root = TrieNode()
        self.currSentence = ""
        #create a trie for all of the given sentences
        for i, sentence in enumerate(sentences):
            self.insert(sentence, times[i])
        #custom less than function for heap to make comparisons according to the frequency and if frequency is same then lexographically
        Pair.__lt__ = lambda self, next: self.sentence > next.sentence if self.count == next.count else self.count < next.count

    def insert(self, sentence, times):
        curr = self.root
        for char in sentence:
            if char not in curr.children:
                curr.children[char] = TrieNode()
            curr = curr.children[char]
            #to the prefix, add the entire sentence and count in it's hashmap
            curr.hashmap[sentence] = curr.hashmap.get(sentence, 0) + times

    def search(self, string):
        curr = self.root
        for char in string:
            if char not in curr.children:
                return {}
            curr = curr.children[char]

        return curr.hashmap

    def input(self, c: str) -> List[str]:
        #if the search input has been completed then add it to the trie
        if c == "#":
            self.insert(self.currSentence, 1)
            self.currSentence = ""
            return []
        else:
            self.currSentence += c
            counts = self.search(self.currSentence)
            heap = []
            #get top three highest frequency sentences for input
            for sent, count in counts.items():
                heapq.heappush(heap, Pair(sent, count))
                if len(heap) > 3:
                    heapq.heappop(heap)

            result = []
            while heap:
                pair = heapq.heappop(heap)
                result.append(pair.sentence)

            return reversed(result)

# Your AutocompleteSystem object will be instantiated and called as such:
# obj = AutocompleteSystem(sentences, times)
# param_1 = obj.input(c)

File: DesignSearchAutoComplete.java
public class DesignSearchAutoComplete {

}

class AutocompleteSystem {
    class TrieNode {
        TrieNode[] children;
        List<String> li;

        public TrieNode() {
            this.children = new TrieNode[256];
            this.li = new ArrayList<>();
        }

    }

    private void insert(String sentences) {
        TrieNode curr = root;
        for (int i = 0; i < sentences.length(); i++) {
            char c = sentences.charAt(i);
            if (curr.children[c - ' '] == null) {
                curr.children[c - ' '] = new TrieNode();
            }
            curr = curr.children[c - ' '];
            curr.li.add(sentences);
        }

    }

    private List<String> search(String prefix) {
        TrieNode curr = root;
        for (int i = 0; i < prefix.length(); i++) {
            char c = prefix.charAt(i);
            if (curr.children[c - ' '] == null) {
                return new ArrayList<>();
            }
            curr = curr.children[c - ' '];
        }
        return curr.li;
    }

    HashMap<String, Integer> map;
    StringBuilder input;
    TrieNode root;

    public AutocompleteSystem(String[] sentences, int[] times) {
        this.map = new HashMap<>();
        this.input = new StringBuilder();
        this.root = new TrieNode();
        for (int i = 0; i < sentences.length; i++) {
            String sentence = sentences[i];
            if (!map.containsKey(sentence)) {
                insert(sentence);
            }
            map.put(sentence, map.getOrDefault(sentence, 0) + times[i]);
        }

    }

    public List<String> input(char c) {
        if (c == '#') {
            String search = input.toString();
            if (!map.containsKey(search)) {
                insert(search);
            }
            map.put(search, map.getOrDefault(search, 0) + 1);
            input = new StringBuilder();
            return new ArrayList<>();
        }
        input.append(c);
        PriorityQueue<String> pq = new PriorityQueue<>((a, b) -> {
            if (map.get(a) == map.get(b)) {
                return b.compareTo(a);

            }
            return map.get(a) - map.get(b);
        });
        String prefix = input.toString();
        List<String> startsWith = search(prefix);
        for (String str : startsWith) {
            pq.add(str);
            if (pq.size() > 3) {
                pq.poll();
            }
        }

        List<String> result = new ArrayList<>();
        while (!pq.isEmpty()) {
            result.add(0, pq.poll());
        }

        return result;

    }

}

